test/cpp/end2end/xds/xds_security_spiffe_end2end_test.cc

```diff
@@ -0,0 +1,477 @@
+          } else {
+            root_cert_info =
+                std::make_shared<RootCertInfo>(it->second.root_certificate);
+          }
+        }
+        if (identity_being_watched) {
+          pem_key_cert_pairs = it->second.identity_key_cert_pairs;
+        }
+        distributor_->SetKeyMaterials(cert_name, std::move(root_cert_info),
+                                      std::move(pem_key_cert_pairs));
+      }
+    });
+  }
+
+  ~FakeCertificateProvider() override {
+    distributor_->SetWatchStatusCallback(nullptr);
+  }
+
+  grpc_core::RefCountedPtr<grpc_tls_certificate_distributor> distributor()
+      const override {
+    return distributor_;
+  }
+
+  grpc_core::UniqueTypeName type() const override {
+    static grpc_core::UniqueTypeName::Factory kFactory("fake");
+    return kFactory.Create();
+  }
+
+ private:
+  int CompareImpl(const grpc_tls_certificate_provider* other) const override {
+    // TODO(yashykt): Maybe do something better here.
+    return grpc_core::QsortCompare(
+        static_cast<const grpc_tls_certificate_provider*>(this), other);
+  }
+
+  grpc_core::RefCountedPtr<grpc_tls_certificate_distributor> distributor_;
+  CertDataMap cert_data_map_;
+};
+
+class FakeCertificateProviderFactory
+    : public grpc_core::CertificateProviderFactory {
+ public:
+  class Config : public grpc_core::CertificateProviderFactory::Config {
+   public:
+    explicit Config(absl::string_view name) : name_(name) {}
+
+    absl::string_view name() const override { return name_; }
+
+    std::string ToString() const override { return "{}"; }
+
+   private:
+    absl::string_view name_;
+  };
+
+  FakeCertificateProviderFactory(
+      absl::string_view name,
+      FakeCertificateProvider::CertDataMapWrapper* cert_data_map)
+      : name_(name), cert_data_map_(cert_data_map) {
+    CHECK_NE(cert_data_map, nullptr);
+  }
+
+  absl::string_view name() const override { return name_; }
+
+  grpc_core::RefCountedPtr<grpc_core::CertificateProviderFactory::Config>
+  CreateCertificateProviderConfig(
+      const grpc_core::Json& /*config_json*/,
+      const grpc_core::JsonArgs& /*args*/,
+      grpc_core::ValidationErrors* /*errors*/) override {
+    return grpc_core::MakeRefCounted<Config>(name_);
+  }
+
+  grpc_core::RefCountedPtr<grpc_tls_certificate_provider>
+  CreateCertificateProvider(
+      grpc_core::RefCountedPtr<grpc_core::CertificateProviderFactory::Config>
+      /*config*/) override {
+    CHECK_NE(cert_data_map_, nullptr);
+    return grpc_core::MakeRefCounted<FakeCertificateProvider>(
+        cert_data_map_->Get());
+  }
+
+ private:
+  absl::string_view name_;
+  FakeCertificateProvider::CertDataMapWrapper* cert_data_map_;
+};
+
+// Global variables for each provider.
+FakeCertificateProvider::CertDataMapWrapper* g_fake1_cert_data_map = nullptr;
+FakeCertificateProvider::CertDataMapWrapper* g_fake2_cert_data_map = nullptr;
+
+//
+// Client-side mTLS tests
+//
+
+class XdsSecurityTest : public XdsEnd2endTest {
```

Comment:
I'm assuming the repetition over here from xds_security_end2end_test.cc is going to go away. And if there's just one test that needs to be added, maybe just add it over there?